spec/index.adoc

```diff
@@ -619,29 +627,97 @@ See <<state-tree>> for details on the state tree.
+Canister methods that do not change the canister state can be executed more efficiently.
+A query call provides that ability: it is a fast, but less secure way to call a canister.
+Only methods that are explicitly marked as “query methods” by the canister can be called this way.
 
-In order to make a query call to canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/query`. The request body consists of an authentication envelope with a `content` map with the following fields:
+In order to make a query call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/query`.
+The request body consists of an authentication envelope with a `content` map with the following fields:
 
-* `request_type` (`text`): Always `query`
+* `request_type` (`text`): Always `"query"`
 * `sender`, `nonce`, `ingress_expiry`: See <<authentication>>
 * `canister_id` (`blob`): The principal of the canister to call.
-* `method_name` (`text`): Name of the canister method to call
-* `arg` (`blob`): Argument to pass to the canister method
+* `method_name` (`text`): Name of the canister method to call.
+* `arg` (`blob`): Argument to pass to the canister method.
 
-If the call resulted in a reply, the response is a CBOR (see <<cbor>>) map with the following fields:
+If the query call resulted in a reply, the response is a CBOR (see <<cbor>>) map with the following fields:
 
-* `status` (`text`): `replied`
+* `status` (`text`): `"replied"`
 * `reply`: a CBOR map with the field `arg` (`blob`) which contains the reply data.
 
 If the call resulted in a reject, the response is a CBOR map with the following fields:
 
-* `status` (`text`): `rejected`
+* `status` (`text`): `"rejected"`
 * `reject_code` (`nat`): The reject code (see <<reject-codes>>).
 * `reject_message` (`text`): a textual diagnostic message.
-* `error_code` (text): an optional implementation-specific textual error code (see <<error-codes>>).
+* `error_code` (`text`): an optional implementation-specific textual error code (see <<error-codes>>).
+
+[#http-signed-query]
+=== Request: Signed query call
+
+NOTE: The signed query call API is considered EXPERIMENTAL. Canister developers must be aware that the API may evolve in a non-backward-compatible way.
+
+A *signed* query call is a query call amended with signatures produced by IC nodes evaluating the query call to certify the query response.
+
+In order to make a signed query call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/signed_query`.
+The request body has the same structure as the request body for <<http-query>> with the `content` map having the following *additional* fields:
+
+* (optional) `return_signatures` (`text`): Determines whether and how many signatures are returned in the response; possible values are `"none"`, `"one"`.
```

Comment:
disclaimer: just a question, I will not force my preferences here :) 

Shouldn't we start with what is needed for the first use case - namely the NNS dapp and then extend the request type as new uses emerge instead of starting with request type that has 4 different combinations.

@derlerd-dfinity

Reply:
It was my initial suggestion to use an enum, yes. But the discussion on this MR revealed that two separate fields are probably better: for replica-signed queries we'd need the variants:
- 1 Signature without certificate
- 1 Signature with certificate
- (Nothing)

For certified queries (which will be what the NNS dapp use case requires) we need to extend this to
- 1 Signature without certificate
- 1 Signature with certificate
- (Nothing)
- Majority of signatures without certificate
- Majority of signatures with certificate

Based on this I think it is better to let the client choose separately for each field instead of forcing a dependency between these two fields and ending up with a huge enum.

We could discuss whether we want to allow the option for `Nothing` (effectively allowing us to drop `return_signatures` _for now_) given that this can also be obtained via `query`. However, this doesn't seem worth it because
- we already know now that we will have to re-introduce `return_signatures` to encode that we want more than one signatures in the immediate follow up feature on certified queries. 
- it gives us the property that `query` and `certified_query` behave the same if no signatures are requested (or the parameters are omitted). So those two endpoints could still be collapsed into one at a later point. 

Neither of these are very strong arguments, but I also think there are no extremely strong arguments for the alternative. So I'd suggest to simply go ahead with this for now in the interest of progress. If we find that this was a mistake we can always revert given that the API is declared experimental.

Reply:
> So those two endpoints could still be collapsed into one at a later point.

(Or just don't introduce a new endpoint if it is merely a fully compatible extension of an existing one…)

Reply:
Do we envision a use case where we want to request a certificate for a query without requesting a signature? If that's not the case, then two independent fields don't make sense to me, and I'd rather use a single, optional field `signature` or similar, which is a record with a boolean (or optional) `with_certificate` inside.

Reply:
I can't think of such a use case at the moment. However, I also don't see a good reason to prevent it. So given that with a single field the number of enum variants would grow quite a bit in the future (at latest when support for aggregating multiple signatures is added; see the [message](https://github.com/dfinity/interface-spec/pull/163#discussion_r1238056069) above) I'd tend to stick with the current proposal.

Reply:
I think we're having different things in mind, I meant that we could have:
```
signed-query-content = {
  ...
  ? signature : signature-options
}

signature-options = {
  with_certificate : bool
  ? multiplicity: "single" / "multiple"
}
```
(names left to taste). This doesn't require huge enums.

Reply:
I see. No strong preference here. I'd leave it up to @mraszyk. If something like this is easy to model and he sees no other issues then it might be an option.

Reply:
> Do we envision a use case where we want to request a certificate for a query without requesting a signature?

The certificate alone could be useful if the caller wants to determine the subnet on which the canister is hosted (e.g., because the caller only wants to talk to a subnet with some feature - fiduciary, EU, ...).

Reply:
Isn't this use case of determining the subnet covered by `read_state` requests?

Reply:
> Isn't this use case of determining the subnet covered by read_state requests?

You can make a read_state request, but you can also make it for a signed_query. The reason for returning the certificate in a signed_query is that you don't need to make an additional read_state request.

Reply:
I'm not convinced by that use case; you'd issue a query to check that the canister is hosted on, say, an EU subnet, but you would be OK not having any evidence that the query was executed correctly. And this would somehow be preferable to issuing a `read_state` request.

So I still think the types I proposed align better with actual use cases, but it seems I'm the only one who thinks that. Given that it's a minor issue, I'll drop it.

Reply:
> you would be OK not having any evidence that the query was executed correctly.

You wouldn't have this evidence in case of a single signature either.

> And this would somehow be preferable to issuing a read_state request.

Issuing a separate read_state request would increase the latency (and slightly also the bandwidth because you would need to again send all the headers etc.).

Reply:
This conversation has been resolved offline.